Add support for put with signal operation

naveen-rn opened the pull request:
This is a work-in-progress proposal to add put with signal operations into the OpenSHMEM standards.
This is related to issue #206. Proposed feature is available as part of Cray  SHMEM implementation.

naveen-rn commented:
Some PENDING ITEMS that needs further discussions:
- add nbi support - can we add as a single proposal?
- for zero size puts - send only the signal is the expected behavior - does Annex C already capture semantics?
- do we need explanation about consecutive puts_signals or put with another put_signal?

nspark commented on content/shmem_put_signal.tex:
```diff
@@ -0,0 +1,83 @@
+\apisummary{
+    The put with signal routines provide a method for copying data from a
+    contiguous local data object to a data object on a specified \ac{PE}
+    and set a remote flag to signal completion.
```
Suggest:
* "put with signal" &#8594; "put-with-signal"
* "...and set a remote flag..." &#8594; "...and subsequently setting a remote flag..."

naveen-rn replied:
resolved - closing outdated changes.

nspark commented on content/shmem_put_signal.tex:
```diff
@@ -0,0 +1,83 @@
+    delivery of its corresponding \dest{} data words into the data object on
+    the remote \ac{PE}. For example, two successive put with signal routines
+    or a successive put followed by a put with signal routine may deliver data
+    out of order.
```
Suggest changing:

> For example, two successive put with signal routines or a successive put followed by a put with signal routine may deliver data out of order.

to something like:

> Without a memory-ordering operation, there is no implied ordering between the delivery of the signal word of a put-with-signal routine and another data transfer. For example, the delivery of the signal word in a sequence consisting of a put routine followed by a put-with-signal routine does not imply delivery of the put routine's data.

naveen-rn replied:
resolved - closing outdated changes.

nspark commented on content/shmem_put_signal.tex:
```diff
@@ -0,0 +1,83 @@
+\apidescription{
+    The routines return after the data has been copied out of the \source{}
+    array on the local \ac{PE}. The delivery of \signal{} flag on the remote
+    \ac{PE} guarantees the delivery of its corresponding \dest{} data words
```
For some reason, the word "guarantees" doesn't quite feel right. I wonder whether something like "indicates" might be better.

naveen-rn replied:
resolved - closing outdated changes.
changed guarantees to indicated.

nspark commented on content/shmem_put_signal.tex:
```diff
@@ -0,0 +1,83 @@
+\apinotes{
+    The \VAR{sig\_addr} data object can be placed in the symmetric data segment
+    or the symmetric heap which can be same or different from the \VAR{dest}
+    data object.
```
As of 1.4, I think we only have one reference to "data segment" (in the threading text, no less), so I'm hesitant to add another without more clearly defining it. In &sect;3.0 (Memory Model), we say:

> In OpenSHMEM the following kinds of data objects are symmetric:
>
> * Global and static C and C++ variables. [...]
>
> [...] Symmetric data objects are stored on the symmetric heap or in the global/static memory section of each PE.

I also think this text could improved with something like:

> The `dest` and `sig_addr` data objects must both be remotely accessible, but may each be allocated from the symmetric heap or global/static memory.

naveen-rn replied:
resolved - removed reference to "data segment" in the explanation.
closing outdated changes.

manjugv commented on content/shmem_put_signal.tex:
```diff
@@ -0,0 +1,85 @@
+\apinotes{
+    The \VAR{dest} and \VAR{sig\_addr} data object must both be remotely
+    accessible, but may each be allocated from the symmetric heap or global/
+    static memory.
```
Should you just call it “symmetric data object” ?

naveen-rn replied:
This was added specifically to note that both dest and sig_addr need not be part of the same SDO type - dest can be from SHEAP and sig_addr from global/static memory. Generalizing it as SDO makes it bit confusing. Let me know, if we reframe it better.

naveen-rn replied:
@manjugv ?

manjugv replied:
Sorry missed it. In that case, my preference is to use two sentences.

naveen-rn replied:
resolved - closing outdated changes.
modified as suggested.

nspark commented on example_code/shmem_put_signal_example.c:
```diff
@@ -0,0 +1,49 @@
+    shmem_barrier_all();
+
+    if (me != 0) {
+        shmem_long_wait_until((long *)&signals[0], SHMEM_CMP_EQ, 1);
```
Feedback from July 23 reading: Use `shmem_wait_until` or `shmem_uint64_wait_until`.

naveen-rn replied:
resolved - we started using C11 generics throughout the example for all SHMEM APIs.
closing outdated changes.

nspark commented on example_code/shmem_put_signal_example.c:
```diff
@@ -0,0 +1,49 @@
+        shmem_long_wait_until((long *)&signals[0], SHMEM_CMP_EQ, 1);
+    }
+
+    shmemx_putmem_signal(data, message, size*sizeof(uint64_t),
```
Feedback from July 23 reading: `shmem_putmem_signal` not `shmemx_*`

naveen-rn replied:
fixed. resolved - closing outdated changes.

nspark commented on example_code/shmem_put_signal_example.c:
```diff
@@ -0,0 +1,49 @@
+
+    uint64_t* message = malloc(size * sizeof(uint64_t));
+    uint64_t* data    = shmem_malloc(size * sizeof(uint64_t));
+    uint64_t* signals = shmem_malloc(sizeof(uint64_t));
```
Feedback from July 23 reading: Might it better emphasize the symmetric heap or static variable distinction to use a static variable here?

naveen-rn replied:
resolved - closing outdated changes.
Example program went through multiple changes and the comments are resolved.

jamesaross commented on example_code/shmem_put_signal_example.c:
```diff
@@ -0,0 +1,49 @@
+
+    free(message);
+    shmem_free(data);
+    shmem_free(signals);
```
shmem_free(signals) can be removed if using `static uint64_t signal = 0;` above.

naveen-rn replied:
resolved - closing outdated changes.
made changes as suggested.

jamesaross commented on example_code/shmem_put_signal_example.c:
```diff
@@ -0,0 +1,49 @@
+#include <stdio.h>
```
Formatting of the example code should be more consistent with other example code.

naveen-rn replied:
Can this be a doc edit and wait till the uniform format is decided from #237

naveen-rn replied:
@jamesaross any comments on this review comment? can i mark this resolved?

naveen-rn replied:
marking as resolved as this change can be added as doc edit or made as part of #237. We dont have any standard formating template.

agrippa commented on content/shmem_put_signal.tex:
```diff
@@ -0,0 +1,85 @@
+    \ac{PE} as the signal. This signal data object must be
+    remotely accessible.}
+    \apiargument{IN}{signal}{Unsigned 64-bit value used to set the remote
+    \VAR{sig\_addr} signal data object.}
```
I found this wording a bit confusing. One of many suggested rewordings:

Unsigned 64-bit value that is assigned to the remote \VAR{sig\_addr} signal data object.

naveen-rn replied:
resolved - modified as per the suggestion. closing outdated changes.

agrippa commented on content/shmem_put_signal.tex:
```diff
@@ -0,0 +1,85 @@
+    ordering between the delivery of the signal word of a put-with-signal
+    routine and another data transfer. For example, the delivery of the signal
+    word in a sequence consisting of a put routine followed by a put-with-signal
+    routine does not imply delivery of the put routine's data.
```
This is probably a question that has been asked and answered in other contexts but: are there any atomicity guarantees on the delivery of the signal? Should there be a note here calling that out? If there are no guarantees, should/could there be a separate set of APIs that deliver the signal atomically (possibly as an atomic operation like AND and not just an atomic set)?

nspark replied:
I think the answers to your questions lie with the resolution of #204, #229, and the overall memory model work. I think one goal of those efforts should be that `shmem_wait_until` (and family) can be safely triggered by atomic operations and the signal of a put-with-signal. If that requires the signal to be an atomic remote write, then so be it.

agrippa replied:
To your knowledge, has there been investigation into something like put-with-atomic-op? Semantically a putmem followed by an atomic operation (e.g. AND).

Is there any benefit in hardware to doing something like that, or would implementations just support it as putmem-fence-atomic?

nspark replied:
The closest I can think of is the "counting puts" proposal from @jdinan & Co. This is equivalent to a put+fence+atomic-add. That is certainly a useful idiom. I'm not sure whether the full generality of put+fence+atomic-op is useful for all operations (i.e., set, add, and, or, xor).

I definitely think there would be application benefit in having hardware support for these idioms. It is often the case that fence (ordering, not necessarily completion) requires completion, due to network API constraints. If that fence can be avoided because the interconnect manages the ordering and atomic op, then the user is introducing fewer latency-sensitive operations.

shamisp commented:
@nspark ordered network will help but if you take in account something like
adaptive routing it may break it. You really want to have the packet small
enough such that you can squeeze it in single MTU.

On Tue, Jul 24, 2018 at 11:11 AM Nick Park <notifications@github.com> wrote:

> *@nspark* commented on this pull request.
> ------------------------------
>
> In content/shmem_put_signal.tex
> <https://github.com/openshmem-org/specification/pull/218#discussion_r204818717>
> :
>
> > +\apireturnvalues{
> +    None.
> +}
> +
> +\apinotes{
> +    The \VAR{dest} and \VAR{sig\_addr} data object must both be remotely
> +    accessible, but may each be allocated from the symmetric heap or global/
> +    static memory.
> +
> +    The delivery of \signal{} flag on the remote \ac{PE} indicates only the
> +    delivery of its corresponding \dest{} data words into the data object on
> +    the remote \ac{PE}. Without a memory-ordering operation, there is no implied
> +    ordering between the delivery of the signal word of a put-with-signal
> +    routine and another data transfer. For example, the delivery of the signal
> +    word in a sequence consisting of a put routine followed by a put-with-signal
> +    routine does not imply delivery of the put routine's data.
>
> The closest I can think of is the "counting puts" proposal from @jdinan
> <https://github.com/jdinan> & Co. This is equivalent to a
> put+fence+atomic-add. That is certainly a useful idiom. I'm not sure
> whether the full generality of put+fence+atomic-op is useful for all
> operations (i.e., set, add, and, or, xor).
>
> I definitely think there would be application benefit in having hardware
> support for these idioms. It is often the case that fence (ordering, not
> necessarily completion) requires completion, due to network API
> constraints. If that fence can be avoided because the interconnect manages
> the ordering and atomic op, then the user is introducing fewer
> latency-sensitive operations.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshmem-org/specification/pull/218#discussion_r204818717>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACIe2MKlfsy53iL2jt63b6zxKk1JCJEjks5uJ0cYgaJpZM4UOpmX>
> .
>

nspark commented:
August F2F discussion:
* @minsii: Should we add clarifications suggesting `shmem_wait_until` or `shmem_test` to clarify portable practices for accessing the signal word?
* @anshumang: Do we explicitly prohibit the signal word from overlapping with the source or destination buffers?

manjugv commented:
August F2F : Ballot postponed.

naveen-rn commented:
@minsii and @shamisp, I suppose a simple reference to put-with-signal (as in 76a5015) in the list of shmem routines which enforces ordering through p2p_sync operations would clarify the portable practice for using put-with-signal. Let me know, If we need some more clarifications.

naveen-rn commented:
Based on discussions with @anshumang  and others - 
@nspark It doesn't look like we could get some form of compiler optimizations from the `restrict` qualifier and also the primary purpose of `restrict` is to guide users. So, I'm not sure whether we need the `restrict` qualifier in the API - can just the details in the description section be sufficient?

naveen-rn commented:
To update, based on RMA WG discussions, we are planning to drop `restrict` qualifier for put-with-signal and can add it when we go through the whole specification. This is because of the following reasons:
1. Adding `restrict` to signal buffer makes sense - but not to both destination and source buffer - that makes it too restrictive
2. Avoid unnecessary compilation warnings

minsii commented:
@naveen-rn Thanks for the clarification fix with point-to-point sync. I have one more question: why do you define `sig_addr` with `uint64_t` type ?

naveen-rn commented:
@minsii In general, I would prefer a fixed width data type for representing the size of the signal, rather than using generic types which varies with platform. If I'm correct the C standards limits the width to 64. Also, the current implementation in Cray hardware supports only 64 bits for signal - hence we matched that with ```uint64_t```. It can be ```int64_t`` - if we prefer the signal to be negative. Are you interested in something different? Say an opaque object like - ```shmem_sig_t``` and restrict the size of this opaque object to 64 bits?

shamisp commented:
Sometimes it is useful to have negative for detection overflow bugs.

Best,
Pasha

On Sun, Sep 23, 2018 at 10:55 Naveen Namashivayam Ravichandrasekaran <
notifications@github.com> wrote:

> @minsii <https://github.com/minsii> In general, I would prefer a fixed
> width data type for representing the size of the signal, rather than using
> generic types which varies with platform. If I'm correct the C standards
> limits the width to 64. Also, the current implementation in Cray hardware
> supports only 64 bits for signal - hence we matched that with uint64_t.
> It can be int64_t`` - if we prefer the signal to be negative. Are you
> interested in something different? Say an opaque object like -shmem_sig_t```
> and restrict the size of this opaque object to 64 bits?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/openshmem-org/specification/pull/218#issuecomment-423826397>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACIe2NwhUuJse82AkJRV_mro057EG2Mpks5ud68HgaJpZM4UOpmX>
> .
>

minsii commented:
I am not sure whether fixed 64bit width is required in spec. Maybe `shmem_sig_t` opaque object is a more portable approach and the implementation can define different types internally ?

jdinan commented:
@nspark If your review has been addressed, could you please dismiss or update your review (ideally before this afternoon's vote on this ticket)?  Thanks!

manjugv commented:
> I am not sure whether fixed 64bit width is required in spec. Maybe `shmem_sig_t` opaque object is a more portable approach and the implementation can define different types internally ?

FWIW, the InfiniBand spec supports only 32bit as immediate data. I'm not sure, if Cray uses similar technique for sending the signal information with the packet on Cray h/w.

shamisp commented:
@manjugv - IBTA I don't think immediate is any useful in this context, regardless of size. It is delivered as part of CQE...

@minsii @naveen-rn if you decide to go with opaque object, you would have to introduce additional API to pull out completion out of the object.

minsii commented:
Can we extend the `shmem_wait*` APIs to support the new opaque object ?

jdinan commented on content/shmem_put_signal.tex:
```diff
@@ -0,0 +1,87 @@
+void @\FuncDecl{shmem\_put\FuncParam{SIZE}\_signal}@(void *dest, const void *source, size_t nelems, uint64_t *restrict sig_addr, uint64_t signal, int pe);
+void @\FuncDecl{shmem\_ctx\_put\FuncParam{SIZE}\_signal}@(shmem_ctx_t ctx, void *dest, const void *source, size_t nelems, uint64_t *restrict sig_addr, uint64_t signal, int pe);
+\end{CsynopsisCol}
+where \SIZE{} is one of \CONST{8, 16, 32, 64, 128}.
```
Given that OpenSHMEM supports the fixed-width integer types, do we want to introduce new SIZE bindings?

naveen-rn replied:
I would prefer the put-with-signal to be consistent with other APIs in the specification. Removing the support for SIZE based will drop uniformity. But, if everyone decides it is necessary - I dont have any strong requirements for SIZE based put-with-signal API.

jdinan replied:
@nspark indicated these API will get used.  We can consider this comment to be resolved.

jdinan commented on content/shmem_put_signal.tex:
```diff
@@ -0,0 +1,87 @@
+\begin{apiarguments}
+    \apiargument{IN}{ctx}{The context on which to perform the operation.
+      When this argument is not provided, the operation is performed on
+      \CONST{SHMEM\_CTX\_DEFAULT}.}
```
Needs to be updated with new text being used for the `ctx` argument.

naveen-rn replied:
Fixed the ctx arg explanation to reflect the recent changes.

naveen-rn replied:
resolved - closing outdated changes.

jdinan commented on content/shmem_put_signal.tex:
```diff
@@ -0,0 +1,87 @@
+    be allocated on the symmetric heap.
+
+    The restrict qualifier in \VAR{sig\_addr} expects the data object to be
+    distinct from \VAR{dest} and \VAR{source} data objects.
```
I don't understand why `sig_add` and `source` can't be the same pointer value corresponding to separate objects that are on different PEs.

jdinan replied:
From 9/24 meeting: Suggestion was to define something like "nonoverlapping buffers" early in the spec so that the definition can also be used in other RMA, AMO, etc. operations.

naveen-rn replied:
resolved - closing outdated changes. removed reference to restrict qualifier.

jdinan commented on content/shmem_put_signal.tex:
```diff
@@ -0,0 +1,87 @@
+    array on the local \ac{PE}. The delivery of \signal{} flag on the remote
+    \ac{PE} indicates the delivery of its corresponding \dest{} data words
+    into the data object on the remote \ac{PE}.
+}
```
The API description appears to be missing a description of what this routine actually does.  This appears to also be an issue with `shmem_put`, but we shouldn't propagate it here.  For example, `shmem_get` states: "The get routines provide a method for copying a contiguous symmetric data object from a different PE to a contiguous data object on the local PE."

naveen-rn replied:
Fixed - duplicated summary into the description

jdinan commented on example_code/shmem_put_signal_example.c:
```diff
@@ -0,0 +1,42 @@
+
+    shmem_init();
+
+    size_t    size    = (2<<10);
```
`2<<10 == 1<<11 == 2048`.  Do you mean 1<<10, for 2^10?

jdinan replied:
I would argue that writing `2 << 10` to get 2048 is poor programming practice and would rather see `1 << 11` or `2048`.

naveen-rn replied:
fixed used 2048 directly

jdinan commented on example_code/shmem_put_signal_example.c:
```diff
@@ -0,0 +1,42 @@
+    int       n       = shmem_n_pes();
+    int       pe      = (me + 1)%n;
+    uint64_t* message = malloc(size * sizeof(uint64_t));
+    uint64_t* data    = shmem_malloc(size * sizeof(uint64_t));
```
Why not use `shmem_calloc`?  You can remove `data[i] = 0` as well as `shmem_barrier_all()` below.

jdinan replied:
I would move the declaration of `sig_addr` above the `shmem_calloc` if you make this change.  I can't convince myself that isn't a race without consulting the C language spec.

naveen-rn replied:
resolved - made changes as per the suggestion. 
closing outdated changes.

jdinan commented on example_code/shmem_put_signal_example.c:
```diff
@@ -0,0 +1,42 @@
+        shmem_uint64_wait_until(&sig_addr, SHMEM_CMP_EQ, 1);
+    }
+
+    shmem_putmem_signal(data, message, size*sizeof(uint64_t),
```
Please use `shmem_put_signal` generic and just `size` as the count argument.

jdinan replied:
The wait_until calls would also need to be updated to use generics.  If you prefer not to use generics, please be consistent and use `shmem_uint64_put_signal` here.

naveen-rn replied:
resolved - used C11 generic SHMEM APIs in the example. closing outdated changes.

jdinan commented on example_code/shmem_put_signal_example.c:
```diff
@@ -0,0 +1,42 @@
+    if (me == 0) {
+        shmem_uint64_wait_until(&sig_addr, SHMEM_CMP_EQ, 1);
+        printf("BCAST with put with signal is complete\n");
+    }
```
The last PE broadcasts to PE 0?  Why?  Also, remove the printf?

naveen-rn replied:
removed this bcast to PE 0 and restored the SHMEM bcast semantics

shamisp commented:
> Can we extend the `shmem_wait*` APIs to support the new opaque object ?

@minsii IMHO a bit hacky. If @naveen-rn decides to take this path, I would suggest to add an API to poll the notification (completions), create the notification object (potentially address or completion queue/CQ ?) and exchange the object identifier (address, CQ #, etc.) between initiator and target. In this way you can abstract an address or some hardware object. My understanding that this was not the goal of the API. The intend was to keep it super simple, aka make 2 puts to two different vaddr ? My $0.02

manjugv commented:
> @manjugv - IBTA I don't think immediate is any useful in this context, regardless of size. It is delivered as part of CQE...

Agree. Talking to Naveen, I was under the impression it was his intention. 

Irrespective, I don't see how both data and signal can be sent in only one packet, at least for 
the IB hardware because it has two virtual addresses. That said, I still support this proposal because we could employ other optimizations.

naveen-rn commented on content/shmem_put_signal.tex:
```diff
@@ -0,0 +1,87 @@
+\begin{apidefinition}
+
+\begin{C11synopsis}
+void @\FuncDecl{shmem\_put\_signal}@(TYPE *dest, const TYPE *source, size_t nelems, uint64_t *restrict sig_addr, uint64_t signal, int pe);
```
I plan to drop ```restrict``` - this was decided only after the last RMA WG meeting - I wasn't sure we can make changes into the PR when it is open for reading/voting. Will change it after today's meeting.

naveen-rn replied:
BTW, ```restrict``` doesn't make sense in this usage.

naveen-rn replied:
removed restrict - closing outdated changes.

naveen-rn commented:
@shamisp @minsii @manjugv - I don't plan to add a new opaque object - it brings lots of other confusions - how can we use it for other RMA/AMO operations, perform memory ordering. I would prefer to stay with any fixed width types - either ```uint64_t``` or ```int64_t``` - If we have strong requirements for negative values in the signal - we can change to ```int64_t```.

shamisp commented:
> @shamisp @minsii @manjugv - I don't plan to add a new opaque object - it brings lots of other confusions - how can we use it for other RMA/AMO operations, perform memory ordering. I would prefer to stay with any fixed width types - either `uint64_t` or `int64_t` - If we have strong requirements for negative values in the signal - we can change to `int64_t`.

👍 IMHO there is a room for a simple API as you proposed and potentially more complicated one that can abstract hardware counters, queues, etc (I think @jdinan  called it counting put or something along the same concept.)

naveen-rn commented:
Voting postponed in Sep specification meeting

naveen-rn commented:
Review comments from Sep specification meeting:

Resolved:
1. Need to update the description with the api summary - not worry about how other RMA operations are described - we can fix them in other PRs
2. No need to remove SIZE based APIs - as we have use cases with structures
3. remove ```restrict``` and fix the description

Pending:
1. Whether we need ```int64_t``` or ```uint64_t```? @nspark was fine with ```uint64_t```. @shamisp suggested the possible use case of negative values (I assume as the signal is not an increment counter - this use case might not be common - but still probable)  - do we need to use union with both ```int64_t``` and ```uint64_t```?

naveen-rn commented:
Based on [RMA WG discussion](https://github.com/openshmem-org/specification/wiki/RMA-WG-09-27-2018) it was clarified that there is no need for change in datatype (no ```int64_t``` or union) required for the put-with-signal proposal. ```uint64_t``` as defined in the current proposal is acceptable

naveen-rn commented:
@nspark @jdinan @shamisp @manjugv If I understand correctly, it was accepted in the previous RMA WG meeting that we will not add restrict qualifier in the API and also we will not have any explanation on the API notes specifically explaining the overlapping semantics for put-with-signal. This is because of the ambiguity in explaining all the scenarios:
1. source/target buffer can overlap 
2. source/signal buffer can overlap 
3. target/signal buffer can overlap

I suppose only 3 is not allowed in put-with-signal and others are undefined. Hence, we decided to drop explaining this particular scenario and decided to deal with this in a separate PR where we explain the general overlapping semantics for all the RMA routines. Let me know, if my understanding is correct?

nspark commented:
@naveen-rn I agree that only Scenario 3 is prohibited. I recall the discussion about the overlap scenarios, but not our decision on when or where to address such prohibitions. I don't recall this affecting other routines, so -- unless we can identify other routines with related overlap restrictions -- I think it should (for now) live with this API listing.

Thinking again about Scenario 3, one _could_ argue that it should still be well defined for the `signal` and `target` buffers to overlap. Since the write of the `target` word(s) are defined to happen before the write of the `signal` word, there actually is (or should be) a well-defined ordering in how memory is updated.

naveen-rn commented:
>  I don't recall this affecting other routines, so -- unless we can identify other routines with related overlap restrictions -- I think it should (for now) live with this API listing.

As the ordering semantics is guaranteed, to me this overlapping restriction in put-with-signal seems to be little forced. We are trying to avoid users from a particular usage, though the implementations are capable of handling it. If there are any hardware issues that we anticipate with the particular target/signal overlapping - I suppose we could then do that. AFAIK, there shouldn't be any.

Also, it would be hard to make a distinct case where we avoid just the target/signal buffer while the other two scenarios are undefined. What happens if both the source and target PEs are the same? It looks we should either address all the overlapping scenarios or we could keep it undefined as we do for all other RMAs. If we plan to address all the overlapping scenarios, then we should make it uniform across all RMAs.

naveen-rn commented:
@nspark @jdinan @shamisp @manjugv 

I would like to add the following statement for handling the overlap semantics:

    The \VAR{sig\_addr} and \VAR{dest} may not be same or overlapping data
    objects.

Or do we need something more detailed:

    The \VAR{sig\_addr} data object is expected to be distinct from \VAR{dest} 
    and \VAR{source} data objects.

minsii commented on content/shmem_put_signal.tex:
```diff
@@ -0,0 +1,91 @@
+    be allocated on the symmetric heap.
+
+    The \VAR{sig\_addr} and \VAR{dest} may not be overlapping in memory.
+
```
Does the term "overlapping" cover both (1) same memory range and (2) not the same but partially overlapping memory range ? I think both cases are illegal here.

The reason I think it might be a little confusing is that, the collectives section uses "overlapping" in "The source and dest arrays may be the *same* array, but they may not be *overlapping* arrays."

Can we say: 
`The \VAR{sig\_addr} and \VAR{dest} may not be the same data object or overlapping in memory.` 

However, I think @jdinan had some concerns about use of "data object".

naveen-rn commented:
Changes made after the October meeting:
https://github.com/openshmem-org/specification/compare/820a4c..1dcf7e

This can be used for unofficial discussions and update announcements during the November 5th specification meeting. 

@nspark @jdinan @jamesaross @agrippa @manjugv  All the open comments based on previous discussions are now outdated. I suppose, every review comments based on these conversations are resolved. Let me know, if we could mark all these conversations as resolved.

manjugv commented:
👍

naveen-rn commented on content/p2p_sync_intro.tex:
```diff
@@ -3,8 +3,9 @@
-the initiator \ac{PE} using the \FUNC{shmem\_fence} and \FUNC{shmem\_quiet}
-routines.
+the initiator \ac{PE} using the put-with-signal(refer
+section~\ref{subsec:shmem_put_signal}, \FUNC{shmem\_fence} and
```
resolved - closing outdated changes.

naveen-rn commented:
Modified and marked many comments as resolved - as these comments were addressed and many files were previously marked by git as outdated.

naveen-rn commented:
Changes made after the October meeting:
https://github.com/openshmem-org/specification/compare/820a4c..1dcf7e

The following changes were made:

1. removed restrict qualifier
2. updated the description section to contain the same explanation from summary
3. updated context argument explanation
4. added better explanation for sig_addr and dest overlap
5. avoided using shmem_malloc and shifted to using shmem_calloc in example
6. changed the size parameter input in example
7. modified all type-based shmem routines to C11-generic routines in the example
8. removed print option from example
9. clarified that the signal update is an atomic operation

We could use these changes for Jan 2019 OpenSHMEM specification meeting.

naveen-rn commented:
Changes made after the October meeting:
https://github.com/openshmem-org/specification/compare/fd54dd3..441d681 

The following changes were made:

1. removed restrict qualifier
2. updated the description section to contain the same explanation from summary
3. updated context argument explanation
4. added better explanation for sig_addr and dest overlap
5. avoided using shmem_malloc and shifted to using shmem_calloc in example
6. changed the size parameter input in example
7. modified all type-based shmem routines to C11-generic routines in the example
8. removed print option from example
9. clarified that the signal update is an atomic operation

We could use these changes for Jan 2019 OpenSHMEM specification meeting.
**[EDIT]**: Updated after rebasing with master

manjugv commented:
January 2019 Meeting: Voting postponed

shamisp commented on content/shmem_put_signal.tex:
```diff
@@ -68,19 +75,21 @@
+    based on the \VAR{sig\_op} signal operator on the remote \ac{PE} must not
+    cause partial updates. Only concurrent accesses on \VAR{sig\_addr} by
+    different signal update operations using the same signal update operator is
+    guaranteed to be exclusive.
```
I would suggest to clarify what "exclusive" means in this context.

manjugv commented:
@naveen-rn @jdinan As I mentioned in the WG meeting, I think it would be useful for defining partial and complete updates. I created an issue to track this and referenced both tickets that gets impacted.

jdinan commented:
@naveen-rn Can this one also be closed?

naveen-rn commented:
Closing this PR as we have opened a new composed PR with both blocking and non-blocking put-with-signal: https://github.com/openshmem-org/specification/pull/275
